lib/base/registry.hpp

```diff
@@ -21,18 +25,27 @@ template<typename U, typename T>
 class Registry
 {
 public:
-	typedef std::map<String, T> ItemMap;
+	typedef std::unordered_map<String, T> ItemMap;
+
+	static Registry* GetInstance()
+	{
+		return Singleton<Registry>::GetInstance();
+	}
 
 	void Register(const String& name, const T& item)
 	{
-		std::unique_lock<std::mutex> lock(m_Mutex);
+		std::unique_lock lock (m_Mutex);
+
+		if (m_Frozen) {
+			BOOST_THROW_EXCEPTION(std::logic_error("Registry is read-only and must not be modified."));
+		}
```

Comment:
You can also perform this check before even locking the mutex. I don't understand this, if you need the mutex for `m_Frozen` as well, why make it atomic?

Reply:
* I did it exactly like #9627 (24b57f0d3a222835178e88489eabd595755ed883) 🤷‍♂️

Reply:
This doesn't answer the questions! And no, you didn’t do exactly the same.
https://github.com/Icinga/icinga2/blob/061338156c71457397f4127a6c9479cdddc9f0e9/lib/base/namespace.cpp#L52-L56
https://github.com/Icinga/icinga2/blob/061338156c71457397f4127a6c9479cdddc9f0e9/lib/base/namespace.cpp#L94-L98
https://github.com/Icinga/icinga2/blob/061338156c71457397f4127a6c9479cdddc9f0e9/lib/base/namespace.cpp#L120-L122

Reply:
Actually, it IS the same: Register/Freeze

1. locks a mutex
2. checks whether instance is frozen

It's critical to synchronize them! Imagine:

1. Register sees not frozen instance
2. Freeze freezes it
3. ReadLockUnlessFrozen doesn't lock it
4. Register modifies it
5. 🤯

Reply:
> Actually, it IS the same

No, it's not!

> It's critical to synchronize them! Imagine:

I don't get it! It's for a reason [atomic](https://en.cppreference.com/w/cpp/atomic/memory_order#Sequentially-consistent_ordering) and if you can't synchronise those without a mutex why make it atomic then?

Reply:
> > Actually, it IS the same
> 
> No, it's not!

We could argue for centuries...

> I don't get it! It's for a reason [atomic](https://en.cppreference.com/w/cpp/atomic/memory_order#Sequentially-consistent_ordering) and if you can't synchronise those without a mutex why make it atomic then?

I can't synchronise **write ops** without a mutex. m_Frozen is atomic to speed up **read ops**.

Reply:
> You can also perform this check before even locking the mutex.

For knowing that I can go on with a write op – yes.

But I also need to ensure nobody freezes, and then reads (unlocked), a registry in the middle of a write operation. (Remember our discussion, exactly like that, about #10414!)